File: Excel/510 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/510 Find Upper, Lower, Numbers & Special Chars Count.xlsx"
input = read_excel(path, range = "A2:A12")
test  = read_excel(path, range = "B2:E12")

result = input %>% 
  mutate(
    Data = ifelse(is.na(Data), "", Data),
    `Upper Case` = str_count(Data, "[A-Z]"),
    `Lower Case` = str_count(Data, "[a-z]"),
    Numbers = str_count(Data, "[0-9]"),
    `Special Chars` = str_count(Data, "[^A-Za-z0-9]")
  ) %>%
  select(-Data) %>%
  mutate(across(everything(), as.numeric))

identical(test, result)
# [1] TRUE

File: Excel/500-599/510/510 Challenge.py
import pandas as pd
import re

path = "510 Find Upper, Lower, Numbers & Special Chars Count.xlsx"
input = pd.read_excel(path, usecols = "A", skiprows = 1)
test = pd.read_excel(path, usecols = "B:E", skiprows = 1)

result = input.copy()
result["Data"] = result["Data"].fillna("")
result["Upper Case"] = result["Data"].apply(lambda x: len(re.findall("[A-Z]", x)))
result["Lower Case"] = result["Data"].apply(lambda x: len(re.findall("[a-z]", x)))
result["Numbers"] = result["Data"].apply(lambda x: len(re.findall("[0-9]", x)))
result["Special Chars"] = result["Data"].apply(lambda x: len(re.findall("[^A-Za-z0-9]", x)))

result = result.drop(columns=["Data"])

print(result.equals(test)) # True